src/paragraph.js

```diff
@@ -38,5 +49,174 @@ export default class Paragraph extends Plugin {
+ *		<table>
+ *			<tr>
+ *				<td>X</td>
+ *				<td>
+ *					<ul>
+ *						<li>Y</li>
+ *						<li>Z</li>
+ *					</ul>
+ *				</td>
+ *			</tr>
+ *		</table>
+ *
+ * Contains five paragraph-like elements – `<h1>` and two `<td>` and two `<li>`.
+ * Hence, if none of the features is going to convert  those elements the above content will be automatically handled
+ * by the paragraph feature and converted to:
+ *
+ *		<p>Foo</p>
+ *		<p>X</p>
+ *		<p>Y</p>
+ *		<p>Z</p>
+ *
+ * Note: The `<td>` containing two `<li>` elements was ignored – the inner-most paragraph-like elements
+ * have priority upon conversion.
+ *
+ * @member {Set.<String>} module:paragraph/paragraph~Paragraph.paragraphLikeElements
+ */
+Paragraph.paragraphLikeElements = new Set( [
+	'p',
+	'h1',
+	'h2',
+	'h3',
+	'h4',
+	'h5',
+	'h6',
+	'td',
+	'li',
+	'div',
+	'dt',
+	'dd',
+	'blockquote'
+] );
+
+const paragraphsToMerge = new WeakSet();
+
+function autoparagraphText( doc, evt, data, consumable, conversionApi ) {
+	// If text wasn't consumed by the default converter...
+	if ( !consumable.test( data.input ) ) {
+		return;
+	}
+
+	// And paragraph is allowed in this context...
+	if ( !doc.schema.check( { name: 'paragraph', inside: data.context } ) ) {
+		return;
+	}
+
+	// Let's do autoparagraphing.
+
+	const paragraph = new ModelElement( 'paragraph' );
+
+	paragraphsToMerge.add( paragraph );
+
+	data.context.push( paragraph );
+
+	const text = conversionApi.convertItem( data.input, consumable, data );
+
+	if ( text ) {
+		data.output = paragraph;
+		paragraph.appendChildren( text );
+	}
+
+	data.context.pop();
+}
+
+function autoparagraphParagraphLikeElements( doc, evt, data, consumable, conversionApi ) {
+	// If this is a paragraph-like element...
+	if ( !Paragraph.paragraphLikeElements.has( data.input.name ) ) {
+		return;
+	}
+
+	// Which wasn't consumed by its own converter...
+	if ( !consumable.test( data.input, { name: true } ) ) {
+		return;
+	}
+
+	// And there are no other paragraph-like elements inside this tree...
+	if ( hasParagraphLikeContent( data.input ) ) {
+		return;
+	}
+
+	// And paragraph is allowed in this context...
+	if ( !doc.schema.check( { name: 'paragraph', inside: data.context } ) ) {
+		return;
+	}
+
+	// Let's convert this element to a paragraph and then all its children.
+
+	consumable.consume( data.input, { name: true } );
+
+	const paragraph = new ModelElement( 'paragraph' );
+
+	data.context.push( paragraph );
+
+	const convertedChildren = conversionApi.convertChildren( data.input, consumable, data );
+
+	paragraph.appendChildren( modelWriter.normalizeNodes( convertedChildren ) );
+
+	// Remove the created paragraph from the stack for other converters.
+	// See https://github.com/ckeditor/ckeditor5-engine/issues/736
+	data.context.pop();
+
+	data.output = paragraph;
+}
+
+// Merges subsequent paragraphs if they should be merged (see shouldMerge).
+function mergeSubsequentParagraphs( evt, data ) {
+	let node;
+
+	if ( isArray( data.output ) ) {
```

Comment:
Hm... I don't know how this applies to this test: https://github.com/ckeditor/ckeditor5-paragraph/pull/12/commits/9a171bf5e2a45bb989e37fa783fe1e97212e694f. It must have been an attribute with a converter there to trigger this case. Previous tests didn't trigger it and they have a lot of unrecognised content and doc frags.

I'm thinking now that this can be caused by the fact that after applying an attribute it there must be 3 text nodes. We have a test for the same input and without bold converter and it worked fine perhaps because there was just one text node (after normalization). But in this case we have 3 and it cannot be normalised so for some reason, it's returned as an array. I think it may be a bug.

I guess we should report a ticket for this in the engine.